File: #113_Path_Sum_II/Solution.java
import java.util.*;

class TreeNode {
	int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int x) { val = x; }
}

public class Solution {
	int target;
	List<List<Integer>> result;
    public List<List<Integer>> pathSum(TreeNode root, int sum) {
        result = new LinkedList();
        if (root == null) return result;
        target = sum;
        LinkedList<Integer> path = new LinkedList();
        // dfs
        dfs(root, 0, path);
        return result;
    }

    private void dfs(TreeNode root, int sum, LinkedList<Integer> path) {
    	path.add(root.val);
    	sum += root.val;
    	if (root.left == null && root.right == null) {
    		if (sum == target) {
    			result.add(new LinkedList(path));
    		}
    	}
    	if (root.left != null) dfs(root.left, sum, path);
    	if (root.right != null) dfs(root.right, sum, path);
    	sum -= path.removeLast();
    }
}

File: #113_Path_Sum_II/Solution.py
class TreeNode:
	def __init__(self, x):
		self.val = x
		self.left = None
		self.right = None

class Solution:
	# @param {TreeNode} root
	# @param {integer} sum
	# @return {integer[][]}
	def pathSum(self, root, sum):
		if not root:
			return []
		re = []
		self.addSum(root, sum, re, [])
		return re

	def addSum(self, root, sum, re, tmp):
		if not root:
			return
		if not root.left and not root.right and root.val == sum:
			re.append(tmp + [root.val]) # list + list
			return
		tmp.append(root.val)
		self.addSum(root.left, sum - root.val, re, tmp)
		self.addSum(root.right, sum - root.val, re, tmp)
		tmp.pop()

sol = Solution()
root = TreeNode(4)
print sol.pathSum(root, 5)